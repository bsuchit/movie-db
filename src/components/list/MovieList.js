import React, { useEffect, useState } from "react";
import PropTypes from 'prop-types';
import _ from 'lodash';
import { Button, Popup, Input } from "semantic-ui-react"; 

import FilterPopup from "components/filter/FilterPopup";
import searchMovies from "utils/utils";
import 'components/components.scss';

import MovieTile from "./MovieTile";
import './MovieList.scss';
import MoviePreviewContainer from "components/preview/MoviePreviewContainer";

const sortByData = [
  {
    name: 'episode_id',
    label: 'Episode'
  },
  {
    name: 'release_date',
    label: 'Release Date'
  }
];

function MovieList({ movies = [] }) {
  const [moviesState, setMoviesState] = useState([]);
  const [searchText, setSearchText] = useState();
  const [sortBySelection, setSortBySelection] = useState();
  const [selection, setSelection] = useState();

  useEffect(() => {
    setMoviesState(movies);
  }, [movies]);

  const handleOnChange = (e) => {
    e.preventDefault();
    setSearchText(e.target.value);
    setMoviesState(searchMovies(movies, e.target.value));
  };

  const handleOnFilterSelected = (name, checked) => {
    setSortBySelection(checked ? name : null);
    setMoviesState(_.orderBy(moviesState, [name], ['asc']));
  };

  const handleOnMovieSelected = (movie) => {
    setSelection(movie);
  };

  return <div>
    <div className="row list_header">
      <Popup 
        content={<FilterPopup
          data={sortByData}
          selection={sortBySelection}
          onSelected={handleOnFilterSelected}
        />} 
        on='click'
        pinned
        trigger={<Button className="sortby_button" content="Sort By" />}
        position="bottom left"
      />
      <Input
        type="text"
        className="search_input"
        placeholder="Type to search..."
        value={searchText}
        onChange={handleOnChange}
        icon='search'
      />
    </div>
    {
      !_.isEmpty(moviesState) ?
    <div className="list_container">
      <div className="list">
        {
            _.map(moviesState, (movie) => movie &&
              <MovieTile
                key={`${movie.title}${movie.episode_id}`}
                data={movie}
                onClick={handleOnMovieSelected}
                selected={selection && selection.episode_id === movie.episode_id}
              />)
        }
      </div>
      {
        selection ?
          <MoviePreviewContainer
            title={selection.title}
            description={selection.opening_crawl}
            episode={selection.episode_id}
            director={selection.director}
            producer={selection.producer}
            releaseDate={selection.release_date}
          />
        :
          <div className="auto_margin">Please select movie to see preview.</div>
      }
      </div>
      :
        <div>No Movies found.</div>
    }
  </div>;
}

MovieList.propTypes = {
  movies: PropTypes.array,
  onSearch: PropTypes.func
};

export default MovieList;
