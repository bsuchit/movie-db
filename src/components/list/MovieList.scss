@import url('https://fonts.googleapis.com/css?family=Montserrat:300,400,700,800');

*{
  box-sizing: border-box;
  margin: 0;
}

html, body{
  margin: 0;
  background: white;
  font-family: 'Montserrat', helvetica, arial, sans-serif;
  font-size: 14px;
  font-weight: 400;
}

.link{
  display: flex;
  text-align: center;
  color: #777;
  text-decoration: none;
  padding: 10px;
}

.movie_card{
  position: relative;
  display: flex;
  height: 100%;
  overflow: hidden;
  border-radius: 10px;
  transition: all 0.4s;
  border-bottom: 1px solid #897f7f;
  border-right: 1px solid #897f7f;
  &:hover{
    transform: scale(1.02);
    box-shadow: 0px 0px 80px -25px rgba(0,0,0, 0.5);
    transition: all 0.4s;
  }
  .info_section{
    position: relative;
    width: 100%;
    height: 100%;
	background-blend-mode: multiply;
    z-index: 0;
    border-radius: 10px;
    .movie_header{
      display: flex;
      flex-direction: row;
      padding: 25px;
      h1{
        color: black;
        font-weight: 400;
      }
      h4{
        color: #555;
        font-weight: 400;
        margin-top: 5px;
      }
      .type{
        color: #0e0e0e;
        margin-top: 5px;
      }
      .locandina{
        position: relative;
        float: left;
        margin-right: 20px;
        height: 120px;
        box-shadow: 0 0 20px -10px rgba(0,0,0,0.5);
      }
    }
  }
}

@media screen and (min-width: 768px) {
  .movie_desc{
    width: 50%;
  }
  
  .info_section{
    background: #f1f1f1;
  }
}

@media screen and (max-width: 768px) {
  .movie_card{
    width: 95%;
    height: auto;
  }

  .movie_header{
    width: 100%;
  }
  
  .info_section{
    background: #f1f1f1;
    display: inline-grid;
  }
}

.vertical {
    border-left: 2px solid #897f7f;
    height: auto;
    margin: 0 20px;
}

.list{
    display: flex;
    max-width: 660px;
    margin: 15px;
    flex-direction: column;
    flex: 0 0 660px;
}

.list_header{
    width: 100%;
    display: flex;
    margin: 15px;
    align-items: stretch;
}

.sortby_button{
    flex: 0 0 100px;
}

.search_input{
    width: calc(100% - 135px);
}

.list_container{
    display: flex;
    flex-direction: row;
    align-items: baseline;
}

.selected{
    background: #a3a3a3;
}

