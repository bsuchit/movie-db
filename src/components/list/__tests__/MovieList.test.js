import React from "react";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import MovieList from "../MovieList";

describe('Given MovieList', () => {
    const movies = [{
        title: 'A New Hope',
        release_date: '1980-12-05',
        episode_id: 1
      },
      {
        title: 'Return of the jedi',
        release_date: '1983-05-25',
        episode_id: 2
    }];

    it('should render correctly with given props', () => {
        const { container } = render(<MovieList movies={movies} />);
        expect(container.firstChild).toMatchSnapshot();
    });

    it('should show search results when input changed', async () => {
        render(<MovieList movies={movies} />);
        const input = screen.getByTestId('search');
        await userEvent.type(input.childNodes[0], 'New');
        let movieTile = screen.getByTestId('A New Hope');
        expect(movieTile).toBeInTheDocument();
        movieTile = screen.queryByTestId('Return of the jedi');
        expect(movieTile).not.toBeInTheDocument();
    });

    it('should sort movies when sort checkbox checked', async () => {
        const { container } = render(<MovieList movies={movies} />);
        const sortButton = screen.getByText('Sort By');
        await userEvent.click(sortButton);
        const checkbox = screen.getByTestId('episode_id');
        await userEvent.click(checkbox.childNodes[0]);
        expect(container.firstChild).toMatchSnapshot();
    });
});
