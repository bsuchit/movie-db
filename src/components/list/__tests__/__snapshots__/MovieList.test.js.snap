// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given MovieList should render correctly with given props 1`] = `
<div>
  <div>
    <div
      class="row list_header"
    >
      <button
        class="ui button sortby_button"
      >
        Sort By
      </button>
      <div
        class="ui icon input search_input"
        data-testid="search"
      >
        <input
          placeholder="Type to search..."
          type="text"
          value=""
        />
        <i
          aria-hidden="true"
          class="search icon"
        />
      </div>
    </div>
  </div>
  <div
    class="list_container"
  >
    <div
      class="list"
    >
      <div
        class="movie_card"
      >
        <div
          class="info_section"
          data-testid="A New Hope"
        >
          <div
            class="movie_header"
          >
            <div
              class="column"
            >
              <h1>
                A New Hope
              </h1>
              <p
                class="type"
              >
                Episode 1
              </p>
            </div>
            <div
              class="vertical"
            />
            <div
              class="column"
            >
              <h4>
                Release Date: 1980-12-05
              </h4>
              <h4>
                Director: undefined
              </h4>
            </div>
          </div>
        </div>
      </div>
      <div
        class="movie_card"
      >
        <div
          class="info_section"
          data-testid="Return of the jedi"
        >
          <div
            class="movie_header"
          >
            <div
              class="column"
            >
              <h1>
                Return of the jedi
              </h1>
              <p
                class="type"
              >
                Episode 2
              </p>
            </div>
            <div
              class="vertical"
            />
            <div
              class="column"
            >
              <h4>
                Release Date: 1983-05-25
              </h4>
              <h4>
                Director: undefined
              </h4>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="auto_margin"
    >
      Please select movie to see preview.
    </div>
  </div>
</div>
`;

exports[`Given MovieList should sort movies when sort checkbox checked 1`] = `
<div>
  <div>
    <div
      class="row list_header"
    >
      <button
        class="ui button sortby_button"
      >
        Sort By
      </button>
      <div
        class="ui icon input search_input"
        data-testid="search"
      >
        <input
          placeholder="Type to search..."
          type="text"
          value=""
        />
        <i
          aria-hidden="true"
          class="search icon"
        />
      </div>
    </div>
  </div>
  <div
    class="list_container"
  >
    <div
      class="list"
    >
      <div
        class="movie_card"
      >
        <div
          class="info_section"
          data-testid="A New Hope"
        >
          <div
            class="movie_header"
          >
            <div
              class="column"
            >
              <h1>
                A New Hope
              </h1>
              <p
                class="type"
              >
                Episode 1
              </p>
            </div>
            <div
              class="vertical"
            />
            <div
              class="column"
            >
              <h4>
                Release Date: 1980-12-05
              </h4>
              <h4>
                Director: undefined
              </h4>
            </div>
          </div>
        </div>
      </div>
      <div
        class="movie_card"
      >
        <div
          class="info_section"
          data-testid="Return of the jedi"
        >
          <div
            class="movie_header"
          >
            <div
              class="column"
            >
              <h1>
                Return of the jedi
              </h1>
              <p
                class="type"
              >
                Episode 2
              </p>
            </div>
            <div
              class="vertical"
            />
            <div
              class="column"
            >
              <h4>
                Release Date: 1983-05-25
              </h4>
              <h4>
                Director: undefined
              </h4>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="auto_margin"
    >
      Please select movie to see preview.
    </div>
  </div>
</div>
`;
